Guide: Code-First Rails

philsturgeon opened the pull request:
Still an early draft, but an attempt at making a full guide that covers how to do code first with rswag, leaving room for design first and traffic sniffing to also get involved. We can do this for all the guides potentially, or split them into different sections. It is going to be rather large...

Samnple code over here: https://github.com/philsturgeon/rails-hello-openapi/
Bump Docs over here: https://bump.sh/philsturgeon/doc/rails-hello-openapi/

ChristopheDujarric commented:
Thanks a lot for your work @philsturgeon ! This is once more a really robust guide.

After reading through it, I'm wondering only about where it best belongs in the Guides site structure.

It does have ties with Bump.sh as our users need a definition file (e.g. OpenAPI file) to publish their API docs. Now it offers much more knowledge than just "here's a walkthrough/cookbook to do this/that with Bump.sh". It provides a lot of guidance to Ruby devs, whether they have a code-first or design-first approach, to get the best out of their OpenAPI spec (and then have it published via Bump.sh).

Hence my recommendation:
* Split the article in two: "Code-first Ruby APIs" and "Design-first Ruby APIs"
* Place them under https://github.com/bump-sh/docs/tree/main/src/_guides/openapi
* Keep a short section in each to explain "that's where you can push this to Bump.sh"

It also feels like we could have [an additional h3 title here](https://github.com/bump-sh/docs/pull/162/files#diff-4d742ad8fb630c987989f0656d5fca98b52084aa46ea77d8ad885227449c3831R173). Something like "Enhancing the docs readability", or so.

It's just my opinion, I'm curious to hear others!

philsturgeon commented:
Great feedback @ChristopheDujarric, I've done exactly that and split the code/design first guides in two and moved them over to the openapi dir.

philsturgeon commented:
Great feedback @ChristopheDujarric, I've done exactly that and split the code/design first guides in two and moved them over to the openapi dir.

ChristopheDujarric commented on src/_guides/openapi/code-first-rails.md:
```diff
@@ -0,0 +1,263 @@
+---
+title: Deploying docs from Rails
```
```suggestion
title: Generating OpenAPI docs for Ruby on Rails with RSwag
```

ChristopheDujarric commented on src/_guides/openapi/code-first-rails.md:
```diff
@@ -0,0 +1,263 @@
+---
+title: Deploying docs from Rails
+authors: phil
+excerpt: Generating OpenAPI for Ruby on Rails APIs with RSwag
```
```suggestion
excerpt: This guide describes, in a code-first approach and RoR codebase, how to generate OpenAPI definition files, enhancing it with contextual information and deploying it to Bump.sh.
```

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,190 @@
+---
+title: Design-first API Rails Development
```
```suggestion
title: Using OpenAPI to simplify building and testing Ruby on Rails APIs
```

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,190 @@
+
+Once you've defined your OpenAPI via [graphical editors](https://openapi.tools/#gui-editors), [text editors](https://openapi.tools/#text-editors), or [traffic sniffing](/guides/openapi/code-first/#traffic-sniffing), you can pop it into your Git repository somewhere like `api/openapi.yaml`. 
+
+Having this raw OpenAPI in your repository means that Bump.sh setup can actually come first, as there is no generation step with API Design first. To set up deployments to Bump.sh you can use the [CLI](https://github.com/bump-sh/cli#bump-deploy-file), [GitHub Actions](https://github.com/marketplace/actions/bump-sh#diff-on-pull-requests-only), or other [Continuous Integration](https://docs.bump.sh ...
```
> means that Bump.sh setup can actually come first

Since the Bump.sh part is now split/moved to the other article, this might need more context.

philsturgeon replied:
Great point, I noticed a few other bits in there which felt lacking and padded them out too. dab6279

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,190 @@
+---
+title: Design-first API Rails Development
+authors: phil
+excerpt: Use OpenAPI to simplify building and testing Ruby on Rails APIs.
```
```suggestion
excerpt: In a Design-first approach and RoR API, leverage OpenAPI to verify specific request responses or write a test suite for your complete API contracts.
```

philsturgeon commented:
@ChristopheDujarric I just noticed this feedback which so far has not been addressed. The link doesnt work now.

> It also feels like we could have [an additional h3 title here](https://github.com/bump-sh/docs/pull/162/files#diff-4d742ad8fb630c987989f0656d5fca98b52084aa46ea77d8ad885227449c3831R173). Something like "Enhancing the docs readability", or so.

If I guessed it would be about talking about Overlays more? I was planning to link it in when the guide is published but that's not been merged yet. https://github.com/bump-sh/docs/pull/153

ChristopheDujarric commented:
> @ChristopheDujarric I just noticed this feedback which so far has not been addressed. The link doesnt work now.
> 
> > It also feels like we could have [an additional h3 title here](https://github.com/bump-sh/docs/pull/162/files#diff-4d742ad8fb630c987989f0656d5fca98b52084aa46ea77d8ad885227449c3831R173). Something like "Enhancing the docs readability", or so.
> 
> If I guessed it would be about talking about Overlays more? I was planning to link it in when the guide is published but that's not been merged yet. #153

A link to the overlay article will be great!

But I was talking about [here](https://github.com/bump-sh/docs/pull/162/files#diff-02cefa32972174db13a031ddcafd0ad44cd5bf65bd8e60e9754b64808db53268R163). It's just a suggestion though.

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,200 @@
+Instead of writing loads of code and sprinkling in some metadata later to create docs, the design-first workflow assumes you create the OpenAPI descriptions before writing any code at all. Once you have the OpenAPI description documents saved in your repository, you can leverage it at every step of the API lifecycle, to produce mock APIs for clients to test assumptions with, build client librarie ...
+
+This guide is going to look at two specific parts of the API design-first workflow that are most helpful to documentation, and show how to set it up in Rails: request validation automatically, and contract testing responses.
+)
```
```suggestion

```

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,200 @@
+
+Alternatively you can grab some sample OpenAPI from the API Guru Marketplace, and click JSON or YAML to download their OpenAPI descriptions.
+
+Either way, one you have some an OpenAPI description document, pop it into your Git repository somewhere like `api/openapi.yaml`. 
```
```suggestion
Either way, once you have some an OpenAPI description document, pop it into your Git repository somewhere like `api/openapi.yaml`. 
```

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,200 @@
+  --doc my-documentation-name \
+  --token my-documentation-token
+
+* Your new documentation version will soon be ready at https://bump.sh/philsturgeon/doc/rails-design-first
```
Do you want to update this one to the new repo under the Bump.sh GitHub Org?

philsturgeon replied:
There's a new Github repo: https://github.com/bump-sh/rails-hello-openapi but that's not yet deploying to any bump docs, and I don't have enough settings to make it so.

I've made my own Bump Examples organization and will make new projects over there, then deploy each of the three sample projects I've made so far.

philsturgeon replied:
Updated links here and elsewhere to point to https://bump.sh/bump-examples/hub/code-samples/

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,200 @@
+
+Either way, one you have some an OpenAPI description document, pop it into your Git repository somewhere like `api/openapi.yaml`. 
+
+Building an API for a bunch of clients is always a tricky one, but by deploying the documentation first you can see if people like the look of the API before you waste a bunch of time building it. Then as you progress through, especially if you are adding these tools to an existing codebase, you will continue to find mistakes in your OpenAPI or your actual API code, improving both as you go until ...
```
```suggestion
Building an API for a bunch of clients is always a tricky one, but by deploying the documentation first you can see if people like the look of the API before you waste a bunch of time building it. Then as you progress through, especially if you are adding these tools to an existing codebase, you will continue to find mistakes in your OpenAPI or your actual API code, improving both as you go until you have a perfect match that will never again be broken, solving the "docs vs code" drift problem, and every fix will be deployed to Bump.sh with each commit/merge.
```

fbraure commented on src/_guides/bump-sh-tutorials/fastapi.md:
```diff
@@ -12,7 +12,8 @@ The following assumes your local machine is configured with Python and FastAPI,
 1. [Create and name](https://bump.sh/docs/new?utm_source=bump&utm_medium=content_hub&utm_campaign=getting_started) your first API documentation. Then, retrieve the name and token of this documentation from the _CI deployment_ settings page.
 
-2. Install the Bump.sh CLI with [npm](https://docs.npmjs.com/cli/v9/configuring-npm/install?v=true) as below, or use [alternative options](/help/continuous-integration/cli), with
+2. Install the Bump.sh CLI with [npm](https://docs.npmjs.com/cli/v9/configuring-npm/install?v=true) as below, or use [alternative options](/help/continuous-integration/cli).
```
The 3. and the 4. are rendered as 1. and 1.
Same same in all guides when previous paragraph finish by a bash. 
I don't remember the tip to solve this. Do you have it @paulRbr ?

philsturgeon replied:
@fbraure I've reverted changes to other guides and will fix them as I work on them to keep collateral damage to a minimum.

philsturgeon commented:
Fixed all feedback and reverted guides other than these Rails guides back to main. I've ogt changes to make to Huma and Fast API but they'll happen in their own PRs to avoid confusion.

ChristopheDujarric commented on src/_guides/openapi/design-first-rails.md:
```diff
@@ -0,0 +1,194 @@
+Instead of writing loads of code and sprinkling in some metadata later to create docs, the design-first workflow assumes you create the OpenAPI descriptions before writing any code at all. Once you have the OpenAPI description documents saved in your repository, you can leverage it at every step of the API lifecycle, to produce mock APIs for clients to test assumptions with, build client librarie ...
+
+This guide is going to look at two specific parts of the API design-first workflow that are most helpful to documentation, and show how to set it up in Rails: request validation automatically, and contract testing responses.
+- [Getting OpenAPI \& Bump.sh Setup](#getting-openapi--bumpsh-setup)
```
```suggestion
- [Getting OpenAPI & Bump.sh Setup](#getting-openapi--bumpsh-setup)
```

scharrier reviewed (commented):
Thanks Phil, it's a great one. LGTM 👍 

It's interesting to see the howto part, I think it would also be super interesting to give some feedback about rswag (like pros/cons), and maybe talk about the alternatives?

Also, yould you just remove the images optimizations commit from this PR and create a dedicated one? (it's just to keep things clean)

Agains, thanks a lot!
